### PR TITLE
feature: single host should allow scheduling to master

### PR DESCRIPTION
**Pull request by hexinatgithub**
Every time I run a Single host cluster, I need to manual remove the Taint.

```shell
$ sealos run labring/kubernetes:v1.25.0 labring/helm:v3.8.2 labring/calico:v3.24.1 --single
# remove taint
$ kubectl taint nodes --all node-role.kubernetes.io/master:NoSchedule-
```

I add a patch to do the untaint automatic #2317.

### CODE REVIEW

**Comment by cuisongliu** [comment]
```
kubeadm init .... --skip-phases=mark-control-plane

---
apiVersion: kubeadm.k8s.io/v1beta3
kind: InitConfiguration
skipPhases:
- mark-control-plane
```

**Comment by hexinatgithub** [comment]
That's sounds great, I will change my implementation.

**Comment by cuisongliu** [comment]
test is ok ?

**Comment by hexinatgithub** [comment]
> test is ok ?

Test tomorrow, may contain some bugs.

**Comment by hexinatgithub** [comment]
Some test case need to add, need to improve.

**Comment by cuisongliu** [comment]
You can refer to the test case in ci and test it with the shell to see if the final single node has stains.

**Comment by cuisongliu** [comment]
https://github.com/labring/sealos/actions/runs/4259381988/jobs/7412043583#step:6:243 same not remove taint ?

**Comment by fengxsong** [comment]
this process should only invoke in `initCluster`, so I would rather implement it in function `pkg/runtime/KubeadmRuntime.generateInitConfigs`. 

```
if sum(len(masters), len(nodes)) == 1 then append skip-phases
```

**Comment by cuisongliu** [comment]
> this process should only invoke in `initCluster`, so I would rather implement it in function `pkg/runtime/KubeadmRuntime.generateInitConfigs`.
> 
> ```
> if sum(len(masters), len(nodes)) == 1 then append skip-phases
> ```

If he just installs the join node behind master0, is this judgment a bit rough?

**Comment by hexinatgithub** [comment]
> 

I think dependency injection will more appropriate

**Comment by hexinatgithub** [comment]
[NewDefaultRuntime](https://github.com/hexinatgithub/sealos/blob/b05ff82d702e9ae757f422b8ff622233a9f25c86/pkg/runtime/runtime.go#L126) is a really confused function signature which provide `*KubeadmConfig` but will not use it.

**Comment by fengxsong** [comment]
> > this process should only invoke in `initCluster`, so I would rather implement it in function `pkg/runtime/KubeadmRuntime.generateInitConfigs`.
> > ```
> > if sum(len(masters), len(nodes)) == 1 then append skip-phases
> > ```
> 
> If he just installs the join node behind master0, is this judgment a bit rough?

once cluster is provisioned, the taints should remove manually.

**Comment by hexinatgithub** [comment]
> > > this process should only invoke in `initCluster`, so I would rather implement it in function `pkg/runtime/KubeadmRuntime.generateInitConfigs`.
> > > ```
> > > if sum(len(masters), len(nodes)) == 1 then append skip-phases
> > > ```
> > 
> > 
> > If he just installs the join node behind master0, is this judgment a bit rough?
> 
> once cluster is provisioned, the taints should remove manually.

It should only skip `mark-control-plane` phase when the cluster is first created and `--single` flag is set.
When user need join more nodes into cluster, he shouldn't provide `--single` flag the first time, the Taint should remain.

**Comment by cuisongliu** [comment]
https://github.com/labring/sealos/actions/runs/4260038410/jobs/7413126020#step:6:250 There should be no stains

**Comment by hexinatgithub** [comment]
> https://github.com/labring/sealos/actions/runs/4260038410/jobs/7413126020#step:6:250 There should be no stains

(CreateProcessor.PreProcess)[https://github.com/hexinatgithub/sealos/blob/b05ff82d702e9ae757f422b8ff622233a9f25c86/pkg/apply/processor/create.go#L87] use [NewDefaultRuntime](https://github.com/hexinatgithub/sealos/blob/b05ff82d702e9ae757f422b8ff622233a9f25c86/pkg/runtime/runtime.go#L126) cause this weird bug.

File: pkg/apply/processor/create.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func (c *CreateProcessor) PreProcess(cluster *v2.Cluster) error {
 	if err := MountClusterImages(cluster, c.Buildah); err != nil {
 		return err
 	}
-	runTime, err := runtime.NewDefaultRuntime(cluster, c.ClusterFile.GetKubeadmConfig())
```
Comment (cuisongliu) [inline pkg/apply/processor/create.go:92]: This logic needs to be merged, and this change may fail the test.

**Comment by cuisongliu** [comment]
> It was accidentally turned off. Please reopen this issue.

你重新提交pr即可

**Comment by hexinatgithub** [comment]
I rewrite my code and pass all test. 
Please check it.

File: .github/workflows/test_run.yml [hunk 2]
```diff
@@ -35,6 +35,50 @@ jobs:
...
+          sudo cat /root/.sealos/default/etc/kubeadm-init.yaml
+          sudo cat /root/.sealos/default/Clusterfile
+      - name: Verify Single Node Cluster Status
+        run: |
+          echo "Verify Cluster"
+          echo "Current system info"
+          sudo /var/lib/sealos/data/default/rootfs/opt/sealctl cri socket
+          sudo /var/lib/sealos/data/default/rootfs/opt/sealctl cri cgroup-driver --short
+          echo "Check Single Node Stains"
+          ! kubectl get nodes -o json | jq ".items[].spec.taints" | grep "control-plane"
```
Comment (cuisongliu) [inline .github/workflows/test_run.yml:74]: It seems that there is no error reported.

File: pkg/clusterfile/clusterfile.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ type Interface interface {
 	GetCluster() *v2.Cluster
 	GetConfigs() []v2.Config
 	GetKubeadmConfig() *runtime.KubeadmConfig
+	SingleMode(bool) error
```
Comment (cuisongliu) [inline pkg/clusterfile/clusterfile.go:46]: SetSingleMode is better ?

**Review by cuisongliu — APPROVED** [review]

**Pull request merged**